File: frontend/types/property.ts
export interface PropertyMetrics {
  noi: number;
  cap_rate: number;
  irr: number;
  cash_on_cash_return: number;
}

export interface CashflowPoint {
  year: number;
  gross_rent: number;
  expenses: number;
  net_cash_flow: number;
}

export interface Property {
  id: string;
  name: string;
  location: string;
  purchase_price: number;
  market_value: number;
  annual_rent: number;
  annual_expenses: number;
  vacancy_rate: number;
  appreciation_rate: number;
  rent_growth_rate: number;
  expense_growth_rate: number;
  hold_years: number;
  metrics: PropertyMetrics;
  cashflow_projection: CashflowPoint[];
}

export interface PropertyCollection {
  properties: Property[];
}

export interface TrendPoint {
  label: string;
  value: number;
}

export interface RiskPoint {
  label: string;
  risk: number;
}

export interface PortfolioAnalytics {
  total_portfolio_value: number;
  average_irr: number;
  cash_on_cash_return: number;
  vacancy_rate: number;
  growth_trend: TrendPoint[];
  risk_profile: RiskPoint[];
}

export interface MonteCarloPathPoint {
  year: number;
  value: number;
  net_cash_flow: number;
}

export interface MonteCarloPath {
  path: MonteCarloPathPoint[];
}

export interface MonteCarloResponse {
  property_id?: string;
  summary: {
    expected_irr: number;
    paths_returned: number;
    iterations_per_property: number;
  };
  distribution: MonteCarloPath[];
}

export interface StressTestResponse {
  property_id?: string;
  stress: {
    avg_stressed_noi: number;
    avg_vacancy: number;
    avg_expense_load: number;
    avg_noi_delta: number;
  };
}

File: backend/models/property.py
"""Property domain models."""
from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, Field


class PropertyBase(BaseModel):
    name: str = Field(..., description="Property name")
    location: str = Field(..., description="City or metro area")
    purchase_price: float = Field(..., gt=0)
    market_value: float = Field(..., gt=0)
    annual_rent: float = Field(..., ge=0)
    annual_expenses: float = Field(..., ge=0)
    vacancy_rate: float = Field(..., ge=0, le=1)
    appreciation_rate: float = Field(0.03, description="Expected annual appreciation")
    rent_growth_rate: float = Field(0.02, description="Expected annual rent growth")
    expense_growth_rate: float = Field(0.015, description="Expected annual expense growth")
    hold_years: int = Field(10, ge=1, le=30)


class PropertyCreate(PropertyBase):
    pass


class PropertyUpdate(BaseModel):
    name: Optional[str]
    location: Optional[str]
    purchase_price: Optional[float]
    market_value: Optional[float]
    annual_rent: Optional[float]
    annual_expenses: Optional[float]
    vacancy_rate: Optional[float]
    appreciation_rate: Optional[float]
    rent_growth_rate: Optional[float]
    expense_growth_rate: Optional[float]
    hold_years: Optional[int]


class CashflowPoint(BaseModel):
    year: int
    gross_rent: float
    expenses: float
    net_cash_flow: float


class MonteCarloPoint(BaseModel):
    year: int
    value: float
    net_cash_flow: float


class MonteCarloPath(BaseModel):
    path: List[MonteCarloPoint]


class PropertyMetrics(BaseModel):
    noi: float
    cap_rate: float
    irr: float
    cash_on_cash_return: float


class PropertyResponse(PropertyBase):
    id: str
    metrics: PropertyMetrics
    cashflow_projection: List[CashflowPoint]


class PropertyCollection(BaseModel):
    properties: List[PropertyResponse]
